src/wallet/wallet_tx_builder.cpp

```diff
@@ -501,38 +501,36 @@ std::pair<uint256, uint256> WalletTxBuilder::SelectOVKs(
 
 PrivacyPolicy TransactionEffects::GetRequiredPrivacyPolicy() const
 {
-    PrivacyPolicy maxPrivacy = PrivacyPolicy::FullPrivacy;
-
-    if (!spendable.orchardNoteMetadata.empty() && payments.HasSaplingRecipient()) {
-        maxPrivacy = PrivacyPolicy::AllowRevealedAmounts;
-    }
-
-    if (!spendable.saplingNoteEntries.empty() && payments.HasOrchardRecipient()) {
-        maxPrivacy = PrivacyPolicy::AllowRevealedAmounts;
-    }
-
-    if (!spendable.sproutNoteEntries.empty() && payments.HasSaplingRecipient()) {
-        maxPrivacy = PrivacyPolicy::AllowRevealedAmounts;
-    }
+    if (!spendable.utxos.empty()) {
+        std::set<std::optional<AddressUFVKMetadata>> uas;
+        for (const auto& utxo : spendable.utxos) {
+            uas.insert(utxo.second);
+        }
 
-    bool hasTransparentSource = !spendable.utxos.empty();
-    if (payments.HasTransparentRecipient()) {
-        if (hasTransparentSource) {
+        if (uas.size() > 1) {
+            if (payments.HasTransparentRecipient()) {
+                return PrivacyPolicy::NoPrivacy;
+            } else {
+                return PrivacyPolicy::AllowLinkingAccountAddresses;
+            }
+        } else if (payments.HasTransparentRecipient()) {
             // TODO: This is the correct policy, but it’s a breaking change from previous behavior,
             //       so enable it separately.
             // maxPrivacy = PrivacyPolicy::AllowFullyTransparent;
-            maxPrivacy = PrivacyPolicy::AllowRevealedSenders;
+            return PrivacyPolicy::AllowRevealedSenders;
         } else {
-            maxPrivacy = PrivacyPolicy::AllowRevealedRecipients;
+            return PrivacyPolicy::AllowRevealedSenders;
         }
-    } else if (hasTransparentSource) {
-        maxPrivacy = PrivacyPolicy::AllowRevealedSenders;
+    } else if (payments.HasTransparentRecipient()) {
+        return PrivacyPolicy::AllowRevealedRecipients;
```

Comment:
This test is now duplicative of line 516?

Reply:
Oh, never mind, this is up a scope.

Reply:
I don’t think so – line 516 only gets tested if we’re spending utxos (thus why it _should_ result in `AllowFullyTransparent`), and this line only gets tested if we’re _not_ spending utxos.

In #6410 (one of the PRs blocked by this one), instead of these cascading / nested conditions, each check is independent and `PrivacyPolicyMeet` is used to resolve to the strongest allowable policy.